src/Reflection/ReflectionClass.php

```diff
@@ -478,40 +493,47 @@ public function getConstructor() : ReflectionMethod
      * Get only the properties for this specific class (i.e. do not search
      * up parent classes etc.)
      *
+     * @param ?int $filter
      * @return ReflectionProperty[]
      */
-    public function getImmediateProperties() : array
+    public function getImmediateProperties(?int $filter = null) : array
     {
-        if (null !== $this->cachedProperties) {
-            return $this->cachedProperties;
-        }
-
-        $properties = [];
-        foreach ($this->node->stmts as $stmt) {
-            if ($stmt instanceof PropertyNode) {
-                $prop = ReflectionProperty::createFromNode($this->reflector, $stmt, $this);
-                $properties[$prop->getName()] = $prop;
+        if (null === $this->cachedProperties) {
+            $properties = [];
+            foreach ($this->node->stmts as $stmt) {
+                if ($stmt instanceof PropertyNode) {
+                    $prop = ReflectionProperty::createFromNode($this->reflector, $stmt, $this);
+                    $properties[$prop->getName()] = $prop;
+                }
             }
+
+            $this->cachedProperties = $properties;
         }
 
-        return $this->cachedProperties = $properties;
+        return array_filter(
+            $this->cachedProperties,
+            function (ReflectionProperty $property) use ($filter) {
+                return null === $filter || $filter & $property->getModifiers();
```

Comment:
Weird that in this scenario you went for the "always true" condition in the filter callback